concordium-contracts-common/src/types.rs

```diff
@@ -448,6 +450,211 @@ pub struct NonZeroThresholdU8<Kind> {
+#[derive(Copy, Clone, Debug, PartialEq, PartialOrd, Eq, Ord)]
+#[repr(transparent)]
+pub struct SignatureEd25519(pub [u8; 64]);
+
+impl fmt::Display for SignatureEd25519 {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        for b in self.0 {
+            write!(f, "{:02x}", b)?;
+        }
+        Ok(())
+    }
+}
+
+impl FromStr for SignatureEd25519 {
+    type Err = ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.len() != 128 {
+            return Err(ParseError {});
+        }
+
+        let mut signature: [u8; 64] = [0u8; 64];
+        for (i, place) in signature.iter_mut().enumerate() {
+            *place = u8::from_str_radix(&s[2 * i..2 * i + 2], 16).map_err(|_| ParseError {})?;
+        }
+
+        Ok(SignatureEd25519(signature))
+    }
+}
+
+/// Signature for a ECDSA (over Secp256k1) message. Must be 64 bytes longs
+/// (serialized in compressed format).
+#[derive(Copy, Clone, Debug, PartialEq, PartialOrd, Eq, Ord)]
+#[repr(transparent)]
+pub struct SignatureEcdsaSecp256k1(pub [u8; 64]);
+
+impl fmt::Display for SignatureEcdsaSecp256k1 {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        for b in self.0 {
+            write!(f, "{:02x}", b)?;
+        }
+        Ok(())
+    }
+}
+
+impl FromStr for SignatureEcdsaSecp256k1 {
+    type Err = ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.len() != 128 {
+            return Err(ParseError {});
+        }
+
+        let mut signature: [u8; 64] = [0u8; 64];
+        for (i, place) in signature.iter_mut().enumerate() {
+            *place = u8::from_str_radix(&s[2 * i..2 * i + 2], 16).map_err(|_| ParseError {})?;
+        }
+
+        Ok(SignatureEcdsaSecp256k1(signature))
+    }
+}
+
+#[cfg(feature = "concordium-quickcheck")]
+/// Arbitrary public keys.
+/// Note that this is a simple generator that might produce an array of bytes
+/// that is not a valid public key.
+impl quickcheck::Arbitrary for PublicKeyEd25519 {
+    fn arbitrary(g: &mut quickcheck::Gen) -> Self {
+        let lower: u128 = quickcheck::Arbitrary::arbitrary(g);
+        let upper: u128 = quickcheck::Arbitrary::arbitrary(g);
+        let mut out = [0u8; 32];
+        out[..16].copy_from_slice(&lower.to_le_bytes());
+        out[16..].copy_from_slice(&upper.to_le_bytes());
+        PublicKeyEd25519(out)
+    }
+}
+
+#[cfg(feature = "std")]
+pub(crate) type KeyIndex = u8;
+
+#[cfg(feature = "std")]
+#[derive(crate::Serialize, Debug, crate::SchemaType, PartialEq, Eq)]
+/// A public indexed by the signature scheme. Currently only a
+/// single scheme is supported, `ed25519`.
+pub enum PublicKey {
+    Ed25519(PublicKeyEd25519),
+}
+
+#[cfg(feature = "std")]
+#[derive(crate::Serialize, Debug, crate::SchemaType, PartialEq, Eq)]
+pub struct CredentialPublicKeys {
+    #[concordium(size_length = 1)]
+    pub keys:      crate::collections::BTreeMap<KeyIndex, PublicKey>,
```

Comment:
```suggestion
    pub keys:      BTreeMap<KeyIndex, PublicKey>,
```

It's already imported in this module (in both std and no-std variants).

Reply:
Unfortunately, the `BTreeMap` is not imported yet, only when `std` feature is explicitly enabled

e.g. this was the import statement before
```
#[cfg(all(feature = "std", feature = "concordium-quickcheck"))]
use std::collections::BTreeMap;
```

Most of the added types now only compile with the `std` feature enabled, e.g.
```
#[cfg(feature = "std"))]
use std::collections::BTreeMap;
```

These new key and signature types are now compiled only when `std` feature is explicitly enabled because we need the `BTreeMap` as well as since we derive `SchemaType` we need `std::{string::String, vec::Vec}` in the crate root.

I acknowledge that this is not all we want (we would like to have these types available also in `no-std` build) but at the same time I want the `SchemaType` to be derived for these types.

Reply:
I pushed changes fixing this now.